src/Workspaces/Remote/ServiceHub.CoreComponents/Microsoft.CodeAnalysis.Remote.ServiceHub.CoreComponents.csproj

```diff
@@ -20,27 +21,106 @@
     <!-- These references need to be deployed to the vsix subfolder containing servicehub bits for .Net Core -->
     <PackageReference Include="Microsoft.VisualStudio.Telemetry" Version="$(MicrosoftVisualStudioTelemetryVersion)" />
     <PackageReference Include="SQLitePCLRaw.bundle_green" Version="$(SQLitePCLRawbundle_greenVersion)" />
-    <!-- We use the content of these two runtime pacakges to exclude BCL library from the dependencies we need to deploy as part of Roslyn vsix. -->
+    <!--
+      We use the content of these two runtime pacakges to: 
+        - exclude BCL library from the dependencies we need to deploy as part of Roslyn vsix.
+        - compile ReadyToRun binaries
+    -->
     <PackageReference Include="Microsoft.NETCore.App.Runtime.win-x64" Version="$(MicrosoftNETCoreAppRuntimewinx64Version)" ExcludeAssets="all" PrivateAssets="all" GeneratePathProperty="true" />
     <PackageReference Include="Microsoft.WindowsDesktop.App.Runtime.win-x64" Version="$(MicrosoftWindowsDesktopAppRuntimewinx64Version)" ExcludeAssets="all" PrivateAssets="all" GeneratePathProperty="true" />
+    <PackageReference Include="Microsoft.NETCore.App.crossgen2.win-x64" Version="$(MicrosoftNETCoreAppcrossgen2winx64Version)" ExcludeAssets="all" PrivateAssets="all" GeneratePathProperty="true" />
   </ItemGroup>
 
-  <Target Name="PublishProjectOutputGroup" DependsOnTargets="Publish" Returns="@(_PublishedFiles)">
+  <Target Name="LocateCrossgenTargets" DependsOnTargets="Publish">
+    <ItemGroup>
+      <!-- Find all Roslyn assemblies that we want to crossgen -->
+      <_R2RAssemblies Include="@(_ResolvedProjectReferencePaths->'%(FileName)%(Extension)')" />
+      <_R2RAssemblies Include="MessagePack.dll" />
+      <!-- Find all assemblies (including Roslyn and all dependencies) from the actual published location -->
+      <_AllPublishedAssemblyPaths Include="$(PublishDir)\**\*.dll" Exclude="$(PublishDir)\**\*.resources.dll" />
+      <_AllPublishedAssemblies Include="@(_AllPublishedAssemblyPaths->'%(FileName)%(Extension)')" >
+        <_FullFilePath>%(FullPath)</_FullFilePath>
+      </_AllPublishedAssemblies>
+      <!-- Get all non-Roslyn assemblies as a subset of all published ones, we don't crossgen those -->
+      <_NoR2RAssemblies Include="@(_AllPublishedAssemblies)" Exclude="@(_R2RAssemblies)" >
+        <_FullFilePath>%(_AllPublishedAssemblies._FullFilePath)</_FullFilePath>
+      </_NoR2RAssemblies>
+      <_NoR2RAssemblyPaths Include="@(_NoR2RAssemblies->'%(_FullFilePath)')" />
+      <!-- Now we get all Roslyn assemblies in the publish folder -->
+      <_R2RAssemblyPaths Include="@(_AllPublishedAssemblyPaths)" Exclude="@(_NoR2RAssemblyPaths)" />
+    </ItemGroup>
+    
+    <PropertyGroup>
+      <CrossgenWorkFolder>$(PublishDir)\CrossGen\</CrossgenWorkFolder>
+      <OriginalAssemblyFolder>$(CrossgenWorkFolder)OriginalAssemblies\</OriginalAssemblyFolder>
+    </PropertyGroup>   
+
+     <Message Text="Moving Files @(_R2RAssemblyPaths)"/>
```

Comment:
>@(_R2RAssemblyPaths) [](http://example.com/codeflow?start=33&length=20)

nit: i'd not include the list in the message. The Move target will log the moved files itself